### PR TITLE
Use STORE for {Class,Function}Def.name in ExpressionContextProvider

### LINKED ISSUE #393: {Class,Function}Def.name considered LOAD by ExpressionContextProvider
The names of class and functions seem more naturally a `STORE` as they introduce a name.

### PR DESCRIPTION
**Pull request by cdonovick**
## Summary
closes #393  by extending ExpressionContextVisitor

## Test Plan
Adds basic tests for the extension

### CODE REVIEW

**Comment by cdonovick** [comment]
Guess I should have ran all tests not just ones I edited.

Seems like this will have more widespread effects than I have anticipated. If you would accept this PR given passing test I would take stab at updating the broken ones.

File: libcst/metadata/expression_context_provider.py [hunk 1]
```diff
@@ -155,6 +155,29 @@ def visit_List(self, node: cst.List) -> Optional[bool]:
...
+        return False
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> Optional[bool]:
+        node.name.visit(
+            ExpressionContextVisitor(self.provider, ExpressionContext.STORE)
+        )
+        node.params.visit(self)
+        node.body.visit(self)
+        for decorator in node.decorators:
+            decorator.visiet(self)
```
Comment (jimmylai) [inline libcst/metadata/expression_context_provider.py:178]: This is a typo.
```suggestion
            decorator.visit(self)
```

**Review by jimmylai — CHANGES_REQUESTED** [review]
This change looks good, just needs to fix the failing tests.
You can run `tox -e autofix` to fix the formatting error in lint CI job.

File: libcst/metadata/tests/test_expression_context_provider.py [hunk 2]
```diff
@@ -376,3 +376,26 @@ def test_for(self) -> None:
                 },
             )
         )
+
+    def test_class(self) -> None:
+        wrapper = MetadataWrapper(parse_module("class Foo: pass"))
```
Comment (jimmylai) [inline libcst/metadata/tests/test_expression_context_provider.py:381]: To make the test comprehensive, let's also test the name of base classes and some names in the class body.

File: libcst/metadata/expression_context_provider.py [hunk 3]
```diff
@@ -155,6 +155,29 @@ def visit_List(self, node: cst.List) -> Optional[bool]:
...
+        node.body.visit(self)
+        for base in node.bases:
+            base.visit(self)
+        for keyword in node.keywords:
+            keyword.visit(self)
+        for decorator in node.decorators:
+            decorator.visit(self)
+        return False
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> Optional[bool]:
```
Comment (jimmylai) [inline libcst/metadata/expression_context_provider.py:171]: The `returns` of a FunctionDef can also have some names.
Let's also visit it and add a test case for it.
https://libcst.readthedocs.io/en/latest/nodes.html#libcst.FunctionDef

**Review by jimmylai — APPROVED** [review]
Looks great!

**Comment by cdonovick** [comment]
@jimmylai Could you create a new release on pypi?  Thanks.

**Comment by jimmylai** [comment]
@zsolt plans to make another release after the pending PRs are merged.
I'm reviewing them now.

**Comment by zsol** [comment]
Wrong Zsolt :) but yes, I expect to make a new release tomorrow.

**Pull request merged**